File: app/src/main/java/com/talktiva/pilot/adapter/AdapterEvent.java
package com.talktiva.pilot.adapter;

import android.app.Activity;
import android.app.Dialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.constraintlayout.widget.ConstraintLayout;
import androidx.recyclerview.widget.RecyclerView;

import com.talktiva.pilot.R;
import com.talktiva.pilot.helper.Utility;
import com.talktiva.pilot.model.Event;
import com.talktiva.pilot.rest.ApiClient;
import com.talktiva.pilot.rest.ApiInterface;

import java.text.SimpleDateFormat;
import java.util.List;
import java.util.Locale;
import java.util.Objects;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class AdapterEvent extends RecyclerView.Adapter<AdapterEvent.EventViewHolder> {

    private ClickListener clickListener;
    private Dialog progressDialog, internetDialog;
    private List<Event> events;
    private Activity activity;
    private Utility utility;
    private int from;

    AdapterEvent(Activity activity, List<Event> events, ClickListener clickListener, int from) {
        this.utility = new Utility(activity);
        this.progressDialog = utility.showProgress();
        this.clickListener = clickListener;
        this.activity = activity;
        this.events = events;
        this.from = from;
    }

    @NonNull
    @Override
    public EventViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int i) {
        return new EventViewHolder(LayoutInflater.from(activity).inflate(R.layout.event_item_child, viewGroup, false));
    }

    @Override
    public void onBindViewHolder(@NonNull final EventViewHolder eventViewHolder, int position) {
        eventViewHolder.bindDataWithViewHolder(events.get(position));
    }

    @Override
    public int getItemCount() {
        return events.size();
    }

    class EventViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.yf_rv_cl)
        ConstraintLayout clItem;

        @BindView(R.id.yf_rv_tv_date)
        TextView tvEventDate;

        @BindView(R.id.yf_rv_tv_title)
        TextView tvTitle;

        @BindView(R.id.yf_rv_tv_full_date)
        TextView tvFullDate;

        @BindView(R.id.yf_rv_tv_add)
        TextView tvAddress;

        @BindView(R.id.yf_rv_iv_private)
        ImageView ivPrivate;

        @BindView(R.id.yf_rv_iv_share)
        ImageView ivShare;

        @BindView(R.id.yf_rv_iv_like)
        ImageView ivLike;

        @BindView(R.id.yf_rv_iv_like_count)
        TextView tvLikeCount;

        @BindView(R.id.yf_rv_iv_more)
        ImageView ivMore;

        EventViewHolder(@NonNull View itemView) {
            super(itemView);
            ButterKnife.bind(this, itemView);
        }

        void bindDataWithViewHolder(Event event) {
            tvEventDate.setTypeface(utility.getFontBold());
            tvTitle.setTypeface(utility.getFontBold());
            tvFullDate.setTypeface(utility.getFontBold());
            tvAddress.setTypeface(utility.getFontRegular());
            tvLikeCount.setTypeface(utility.getFontBold());

            tvEventDate.setText(new SimpleDateFormat("MMM", Locale.US).format(event.getEventDate()).concat("\n").concat(new SimpleDateFormat("dd", Locale.US).format(event.getEventDate())));
            tvTitle.setText(event.getTitle());

            //noinspection deprecation
            tvFullDate.setText(event.getEventDate().toLocaleString());
            tvAddress.setText(event.getCreatorFirstName().concat(" ").concat(event.getCreatorLasttName().concat(" | ")).concat(event.getLocation()));
            tvLikeCount.setText(String.valueOf(event.getLikeCount()));

            if (event.isPrivate()) {
                ivPrivate.setVisibility(View.VISIBLE);
            } else {
                ivPrivate.setVisibility(View.GONE);
            }

            if (event.canInviteGuests()) {
                ivShare.setVisibility(View.VISIBLE);
            } else {
                ivShare.setVisibility(View.GONE);
            }

            if (event.isHasLiked()) {
                ivLike.setImageResource(R.drawable.ic_liked);
                ivLike.setTag("1");
            } else {
                ivLike.setImageResource(R.drawable.ic_like);
                ivLike.setTag("0");
            }

            switch (from) {
                case 0:
                    ivMore.setVisibility(View.GONE);
                    break;
                case 1:
                    ivMore.setVisibility(View.VISIBLE);
                    break;
                case 2:
                    ivMore.setVisibility(View.VISIBLE);
                    break;
            }

            clItem.setOnClickListener(v -> clickListener.onPositionClicked(clItem, events.get(getAdapterPosition()), from));

            ivShare.setOnClickListener(v -> clickListener.onPositionClicked(ivShare, events.get(getAdapterPosition()), from));

            ivMore.setOnClickListener(v -> clickListener.onPositionClicked(ivMore, events.get(getAdapterPosition()), from));

            ivLike.setOnClickListener(v -> {
                if (v.getTag().toString().equalsIgnoreCase("0")) {
                    ApiInterface apiInterface = ApiClient.getClient().create(ApiInterface.class);
                    Call<Event> call = apiInterface.likeEvent(activity.getResources().getString(R.string.token_prefix).concat(" ").concat(activity.getResources().getString(R.string.token_amit)), event.getEventId());
                    call.enqueue(new Callback<Event>() {
                        @Override
                        public void onResponse(@NonNull Call<Event> call, @NonNull Response<Event> response) {
                            if (response.isSuccessful()) {
                                if (Objects.requireNonNull(response.body()).isHasLiked()) {
                                    ivLike.setImageResource(R.drawable.ic_liked);
                                    ivLike.setTag("1");
                                    tvLikeCount.setText(String.valueOf(response.body().getLikeCount()));
                                }
                            } else {
                                if (response.code() >= 300 && response.code() < 500) {
                                    utility.showMsg(response.message());
                                } else if (response.code() >= 500) {
                                    internetDialog = utility.showError(activity.getResources().getString(R.string.server_msg), activity.getResources().getString(R.string.dd_try), v -> utility.dismissDialog(internetDialog));
                                    internetDialog.show();
                                }
                            }
                        }

                        @Override
                        public void onFailure(@NonNull Call<Event> call, @NonNull Throwable t) {
                            if (t.getMessage().equalsIgnoreCase("timeout")) {
                                internetDialog = utility.showError(activity.getResources().getString(R.string.time_out_msg), activity.getResources().getString(R.string.dd_ok), v -> utility.dismissDialog(internetDialog));
                                internetDialog.show();
                            }
                        }
                    });
                }
            });
        }
    }
}

File: app/src/main/java/com/talktiva/pilot/adapter/AdapterEvent.kt
package com.talktiva.pilot.adapter

import android.app.Dialog
import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import androidx.recyclerview.widget.RecyclerView
import com.google.gson.Gson
import com.google.gson.reflect.TypeToken
import com.talktiva.pilot.R
import com.talktiva.pilot.helper.AppConstant
import com.talktiva.pilot.helper.Utility
import com.talktiva.pilot.model.Event
import com.talktiva.pilot.rest.ApiClient
import com.talktiva.pilot.rest.ApiInterface
import com.talktiva.pilot.results.ResultError
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException
import java.text.DateFormat
import java.text.SimpleDateFormat
import java.util.*

class AdapterEvent internal constructor(private val context: Context, private val events: List<Event>, private val clickListener: ClickListener, private val from: Int) : RecyclerView.Adapter<AdapterEvent.EventViewHolder>() {

    private var internetDialog: Dialog? = null

    override fun onCreateViewHolder(viewGroup: ViewGroup, i: Int): EventViewHolder {
        return EventViewHolder(LayoutInflater.from(context).inflate(R.layout.item_child, viewGroup, false))
    }

    override fun onBindViewHolder(eventViewHolder: EventViewHolder, position: Int) {
        eventViewHolder.bindDataWithViewHolder(events[position])
    }

    override fun getItemCount(): Int {
        return events.size
    }

    inner class EventViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private var clItem = itemView.findViewById<ConstraintLayout>(R.id.yf_rv_cl)
        private var tvEventDate = itemView.findViewById<TextView>(R.id.yf_rv_tv_date)
        private var tvTitle = itemView.findViewById<TextView>(R.id.yf_rv_tv_title)
        private var tvFullDate = itemView.findViewById<TextView>(R.id.yf_rv_tv_full_date)
        private var tvAddress = itemView.findViewById<TextView>(R.id.yf_rv_tv_add)
        private var ivPrivate = itemView.findViewById<ImageView>(R.id.yf_rv_iv_private)
        private var ivShare = itemView.findViewById<ImageView>(R.id.yf_rv_iv_share)
        private var ivLike = itemView.findViewById<ImageView>(R.id.yf_rv_iv_like)
        private var tvLikeCount = itemView.findViewById<TextView>(R.id.yf_rv_iv_like_count)
        private var ivMore = itemView.findViewById<ImageView>(R.id.yf_rv_iv_more)

        fun bindDataWithViewHolder(event: Event) {
            tvEventDate.typeface = Utility.fontBold
            tvTitle.typeface = Utility.fontBold
            tvFullDate.typeface = Utility.fontBold
            tvAddress.typeface = Utility.fontRegular
            tvLikeCount.typeface = Utility.fontBold

            tvEventDate.text = SimpleDateFormat("MMM", Locale.US).format(event.eventDate).plus("\n").plus(SimpleDateFormat("dd", Locale.US).format(event.eventDate))
            tvTitle.text = event.title

            val dateFormat = DateFormat.getDateTimeInstance(DateFormat.DEFAULT, DateFormat.SHORT)
            tvFullDate.text = dateFormat.format(event.eventDate)
            tvAddress.text = event.creatorFullName.plus(" | ").plus(event.location)
            tvLikeCount.text = event.likeCount.toString()

            if (event.isPrivate!!) {
                ivPrivate.visibility = View.VISIBLE
            } else {
                ivPrivate.visibility = View.GONE
            }

            if (event.canInviteGuests!!) {
                ivShare.visibility = View.VISIBLE
            } else {
                ivShare.visibility = View.GONE
            }

            if (event.isHasLiked!!) {
                ivLike.setImageResource(R.drawable.ic_liked)
                ivLike.tag = "1"
            } else {
                ivLike.setImageResource(R.drawable.ic_like)
                ivLike.tag = "0"
            }

            when (from) {
                0 -> ivMore.visibility = View.GONE
                1 -> ivMore.visibility = View.VISIBLE
                2 -> ivMore.visibility = View.VISIBLE
            }

            clItem.setOnClickListener { clickListener.onPositionClicked(clItem!!, events[adapterPosition], from) }

            ivShare.setOnClickListener { clickListener.onPositionClicked(ivShare!!, events[adapterPosition], from) }

            ivMore.setOnClickListener { clickListener.onPositionClicked(ivMore!!, events[adapterPosition], from) }

            ivLike.setOnClickListener { v ->
                if (v.tag.toString().equals("0", ignoreCase = true)) {
                    if (Utility.isConnectingToInternet) {
                        val apiInterface = ApiClient.client.create(ApiInterface::class.java)
                        val call = apiInterface.likeEvent(Utility.getPreference(AppConstant.PREF_T_TYPE).plus(" ").plus(Utility.getPreference(AppConstant.PREF_A_TOKEN)), event.eventId)
                        call.enqueue(object : Callback<Event> {
                            override fun onResponse(call: Call<Event>, response: Response<Event>) {
                                if (response.isSuccessful) {
                                    if (response.body()!!.isHasLiked!!) {
                                        ivLike.setImageResource(R.drawable.ic_liked)
                                        ivLike.tag = "1"
                                        tvLikeCount.text = response.body()!!.likeCount.toString()
                                    }
                                } else {
                                    try {
                                        val resultError = Gson().fromJson<ResultError>(Objects.requireNonNull(response.errorBody())!!.string(), object : TypeToken<ResultError>() {
                                        }.type)
                                        internetDialog = Utility.showAlert(context, resultError.errorDescription, true, View.VISIBLE, R.string.dd_try, View.OnClickListener { Utility.dismissDialog(internetDialog) }, View.GONE, null, null)
                                        internetDialog!!.show()
                                    } catch (e: IOException) {
                                        e.printStackTrace()
                                    }

                                }
                            }

                            override fun onFailure(call: Call<Event>, t: Throwable) {
                                if (t.message.equals("timeout", ignoreCase = true)) {
                                    internetDialog = Utility.showError(context, R.string.time_out_msg, R.string.dd_ok, View.OnClickListener { Utility.dismissDialog(internetDialog) })
                                    internetDialog!!.show()
                                }
                            }
                        })
                    }
                }
            }
        }
    }
}
